snetd/cmd/components.go

```diff
@@ -78,119 +64,144 @@ func isFileExist(fileName string) bool {
-		}
-		err = etcdServer.Start()
-		if err != nil {
-			return errors.Wrap(err, "error during etcd server starting")
-		}
-		components.etcdServer = etcdServer
+	if !enabled {
+		return nil
 	}
-	return
-}
 
-func (components *Components) initPaymentChannelStorage() (err error) {
-	var delegateStorage escrow.AtomicStorage
-	if config.GetString(config.PaymentChannelStorageTypeKey) == "etcd" {
-		client, err := etcddb.NewEtcdClient()
-		if err != nil {
-			return errors.Wrap(err, "unable to create etcd client")
-		}
-		components.etcdClient = client
-		delegateStorage = client
-	} else {
-		delegateStorage = escrow.NewMemStorage()
+	server, err := etcddb.GetEtcdServer()
+	if err != nil {
+		log.WithError(err).Panic("error during etcd config parsing")
 	}
 
-	components.paymentChannelStorage = escrow.NewCombinedStorage(
-		components.blockchain,
-		escrow.NewPaymentChannelStorage(delegateStorage),
-	)
+	err = server.Start()
+	if err != nil {
+		log.WithError(err).Panic("error during etcd server starting")
+	}
 
-	return
+	components.etcdServer = server
+	return server
 }
 
-func (components *Components) initGrpcInterceptor() (err error) {
-	if !components.blockchain.Enabled() {
-		log.Info("Blockchain is disabled: no payment validation")
-		components.grpcInterceptor = handler.NoOpInterceptor
-		return nil
+func (components *Components) EtcdClient() *etcddb.EtcdClient {
+	if components.etcdClient != nil {
+		return components.etcdClient
 	}
 
-	log.Info("Blockchain is enabled: instantiate payment validation interceptor")
-	components.grpcInterceptor = handler.GrpcStreamInterceptor(
-		blockchain.NewJobPaymentHandler(components.blockchain),
-		escrow.NewEscrowPaymentHandler(
-			components.blockchain,
-			components.paymentChannelStorage,
-			escrow.NewIncomeValidator(components.blockchain),
-		),
-	)
-	return nil
-}
+	// start etcd server if enabled as client will try connecting to it
+	_ = components.EtcdServer()
 
-func (components *Components) initPaymentChannelStateService() (err error) {
-	components.paymentChannelStateService = escrow.NewPaymentChannelStateService(components.PaymentChannelStorage())
-	return nil
+	client, err := etcddb.NewEtcdClient()
+	if err != nil {
+		log.WithError(err).Panic("unable to create etcd client")
+	}
+
+	components.etcdClient = client
+	return components.etcdClient
 }
 
-func (components *Components) Close() {
-	if components.db != nil {
-		components.db.Close()
+func (components *Components) PaymentChannelStorage() escrow.PaymentChannelStorage {
+	if components.paymentChannelStorage != nil {
+		return components.paymentChannelStorage
 	}
-	if components.etcdClient != nil {
-		components.etcdClient.Close()
+
+	var delegateStorage escrow.AtomicStorage
+	if config.GetString(config.PaymentChannelStorageTypeKey) == "etcd" {
+		delegateStorage = components.EtcdClient()
+	} else {
+		delegateStorage = escrow.NewMemStorage()
 	}
 	if components.etcdServer != nil {
```

Comment:
Probably some conflicts during merging.
EtcdServer shouldn't be closed after PaymentChannelStorage creation.

Reply:
Thanks!